src/devices/Mcp3xxx/Mcp3Base.cs

```diff
@@ -0,0 +1,169 @@
+            Differential = 1,
+            ///<summary>The value is the difference in voltage between two pins with the second pin being the positive one</summary>
+            InvertedDifferential = 2
+        }
+
+        private SpiDevice _spiDevice;
+
+        /// <summary>
+        /// Constructs Mcp3Base instance
+        /// </summary>
+        /// <param name="spiDevice">Device used for SPI communication</param>
+        public Mcp3Base(SpiDevice spiDevice)
+        {
+            _spiDevice = spiDevice;
+        }
+
+        /// <summary>
+        /// Disposes Mcp3Base instances
+        /// </summary>
+        public void Dispose()
+        {
+            _spiDevice?.Dispose();
+            _spiDevice = null;
+        }
+
+        /// <summary>
+        /// Reads a value from the device
+        /// </summary>
+        /// <param name="adcRequest">A bit pattern to be sent to the ADC.</param>
+        /// <param name="adcRequestLengthBytes">The number of bytes to be sent to the ADC containing the request and returning the response.</param>
+        /// <param name="adcResolutionBits">The number of bits in the returned value</param>
+        /// <param name="delayBits">The number of bits to be delayed between the request and the response being read.</param>
+        /// <returns>A value corresponding to a voltage level on the input pin described by the request.</returns>
+        protected int ReadInternal(int adcRequest, int adcRequestLengthBytes, int adcResolutionBits, int delayBits)
+        {
+            int retval = 0;
+
+            Span<byte> requestBuffer = stackalloc byte[adcRequestLengthBytes];
+            Span<byte> responseBuffer = stackalloc byte[adcRequestLengthBytes];
+
+            // shift the request left to make space in the response for the number of bits in the 
+            // response plus the conversion delay and plus 1 for a null bit.
+            adcRequest <<= (adcResolutionBits + delayBits + 1);
+
+            // take the resuest and put it in a byte array
+            for (int i = 0; i < requestBuffer.Length; i++)
+            {
+                requestBuffer[i] = (byte)(adcRequest >> (adcRequestLengthBytes - i - 1) * 8);
+            }
+
+            _spiDevice.TransferFullDuplex(requestBuffer, responseBuffer);
+
+            // transfer the response from the ADC into the return value
+            for (int i = 0; i < responseBuffer.Length; i++)
+            {
+                retval <<= 8;
+                retval += responseBuffer[i];
+            }
+
+            // test the response from the ADC to check that the null bit is actually 0
+            if ((retval & (1 << adcResolutionBits)) != 0)
+            {
+                throw new InvalidOperationException("Invalid data was read from the sensor");
+            }
+
+            // return the ADC response with any possible higer bits masked out
+            return retval & (int)(0xFFFFFFFF >> (32 - adcResolutionBits));
```

Comment:
I'm unclear if shifting `0xFFFFFFFF` will work as expected - AFAIR signed integers will preserve sign when shifting (so essentially first bit will stay 1) - might be worth validating this or starting with uint and then casting to int

Reply:
assuming this is correct please make types more explicit (i.e. explicit cast) or going reverse way `((1 << adcResolutionBits) - 1)` (might be +/- 1 somewhere)

Reply:
It actually works I think because 0xFFFFFFFF cannot be cast to an int and thus the compiler assumes a uint. perhaps to be clear I should use 0xFFFFFFFFU.

Anyway I do like your method but it also has an issue if you ask for 32 bits of resolution as it would return 0. i.e. 1 << 32 == 0. I think 1L << 32 == 0xFFFFFFFF so i'll use that